### PR TITLE
Adding documentation

### PR DESCRIPTION
**Pull request by pearj**
Adding documentation to the idm functions.

Copying the documentation from the ForgeRock docs onto the openidm functions as well as the typescript wrappers.

This is still in progress.

### CODE REVIEW

**Comment by pearj** [comment]
@dspasojevic do you think I'm adding too much documentation on some of the functions? Especially the patch one? Or is the more doco the better?

**Comment by dspasojevic** [comment]
Generally, I reckon the more documentation the better. Adding a link from the main README to the appropriate backstage docs may be a useful way for new users to orient themselves.

**Review by dspasojevic — APPROVED** [review]

**Comment by pearj** [comment]
Thanks @dspasojevic, I was going to update the README.md as well to give a general overview of the primary features. But links to the ForgeRock docs do make sense too.

**Comment by pearj** [comment]
Hey @dspasojevic, are you ok with this being merged? I've added a fairly detailed readme complete with animations. There are a few gaps in the tsdoc, as I've run out of time. But the readme is fairly complete and will be infinitely more useful than it was previously.

File: README.md [hunk 1]
```diff
@@ -11,8 +11,599 @@
...
+### Supported Scenarios
+
+Code generation supports the following scenarios:
+
+| Scenario | Notes |
+| -------- | ----- |
+| Basic types | `string`, `number`/`integer`, `boolean`, `object`, `array` |
+| [Complex Objects](#complex-objects) | Objects that have defined properties, are generated as separate sub-types, and have unlimited levels of nesting, more info [below](#complex-objects). |
+| [Relationships](#relationships) | Fields that are [relationships](#relationships) simply use the target relationship as the type wrapped in a `ReferenceType`. It gracefully degrades when it can't find the target type to the `Record` type, this usually happens when referencing `internal` types which are not present in `managed.json`. |
+| Nullable fields | Fields marked as _nullable_ in the schema have add `null` as a Typescript [union type](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types), e.g. `string \| null`. |
```
Comment (dspasojevic) [inline README.md:115]: nit: "have add `null` as a" -> "are unioned with `null`" ?

File: README.md [hunk 2]
```diff
@@ -11,8 +11,599 @@
...
+}
+```
+
+Which defines `_grantType`, but the code generation won't create a type with `_grantType` in it. Perhaps in the future this scenario will be supported if there is interest.
+
+Additionally, when using relationships and specifying which fields should be returned isn't supported when using [automatic type narrowing](#automatic-type-narrowing) as the type narrowing doesn't support forward slashes (eg `manager/givenName`), more information is at [field name limitations](#field-name-limitations).
+
+## Type-safe Wrapper Functions
+
+During code generation an addition object called `idm` is generated. The `idm` object is an extension of the built-in `openidm` object, but with wrappers added for all `managed` objects and connectors (`system`) objects.
```
Comment (dspasojevic) [inline README.md:287]: nit: addition -> additional?

File: README.md [hunk 3]
```diff
@@ -11,8 +11,599 @@
...
+
+These wrapper functions are where the power of the types really shines. As seen in the animation below, VS Code is able to auto-complete managed objects and connectors, as well as showing the fields available including any description/title.
```
Comment (dspasojevic) [inline README.md:289]: future: Parts of this (down to "There is one additional function") reads like the hook that should be at the top of the README. Maybe consider a smaller section explaining who should want to use this library to the top of the README.

**Comment by dspasojevic** [comment]
LGTM, though there are a couple of small grammatical issues that could be cleaned up.

File: README.md (hunk 3 above)
Comment (pearj) [reply README.md:289]: Yes you are right, bits of the "Type-safe wrapper functions" section do summarise the use case pretty well, having the animation near the top so people can see straight away why it's useful is probably a good idea too.

I'll see if I can add something tonight,

File: README.md (hunk 1 above)
Comment (pearj) [reply README.md:115]: Thanks for picking that up

File: README.md (hunk 2 above)
Comment (pearj) [reply README.md:287]: Thanks, I'll fix that

**Pull request merged**